File: 985 - Sum of Even Numbers After Queries.cpp
#include <vector>
#include <unordered_set>

using namespace std;

class Solution {
public:
    vector<int> sumEvenAfterQueries(vector<int>& nums, vector<vector<int>>& queries) {
        unordered_set<int> evenIdx;
        int evenSum = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] % 2 == 0) {
                evenSum += nums[i];
                evenIdx.insert(i);
            }
        }
        vector<int> res(queries.size());
        for (int j = 0; j < queries.size(); j++) {
            int i = queries[j][1], val = queries[j][0];
            if ((nums[i] + val) % 2 == 0) {
                evenSum += val;
                if (evenIdx.find(i) == evenIdx.end()) {
                    evenIdx.insert(i);
                    evenSum += nums[i];
                }
            } else if (evenIdx.find(i) != evenIdx.end()) {
                evenSum -= nums[i];
                evenIdx.erase(i);
            }
            nums[i] += val;
            res[j] = evenSum;
        }
        return res;
    }
};

File: 985 - Sum of Even Numbers After Queries.js
/**
 * @param {number[]} nums
 * @param {number[][]} queries
 * @return {number[]}
 */
var sumEvenAfterQueries = function (nums, queries) {
    let evenIdx = new Set();
    let evenSum = 0;
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] % 2 == 0) {
            evenSum += nums[i];
            evenIdx.add(i);
        }
    }
    let res = new Array(queries.length);
    for (let j = 0; j < queries.length; j++) {
        let i = queries[j][1], val = queries[j][0];
        if ((nums[i] + val) % 2 == 0) {
            evenSum += val;
            if (!evenIdx.has(i)) {
                evenSum += nums[i];
                evenIdx.add(i);
            }
        } else if (evenIdx.has(i)) {
            evenIdx.delete(i);
            evenSum -= nums[i];
        }
        nums[i] += val;
        res[j] = evenSum;
    }
    return res;
};


let a = [1,2,3,4];
let b = [[1,0],[-3,1],[-4,0],[2,3]];
sumEvenAfterQueries(a, b);